### PR TITLE
upgrade TypeScript to latest

### PR DESCRIPTION
**Pull request by paul-marechal**
Upgrade TypeScript version and fix new compilation errors.

#### How to test

Everything should work like before.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/source-tree/source-tree.ts [hunk 1]
```diff
@@ -41,50 +43,56 @@ export class SourceTree extends TreeImpl {
...
+            existing.element = element;
+            existing.parent = parent;
+        }
         if (CompositeTreeElement.hasElements(element)) {
-            if (updated) {
-                if (!ExpandableTreeNode.is(updated)) {
-                    Object.assign(updated, { expanded: false });
+            if (existing) {
+                if (!ExpandableTreeNode.is(existing)) {
+                    (existing as any).expanded = false;
```
Comment (colin-grant-work) [inline packages/core/src/browser/source-tree/source-tree.ts]: TS seems happy with these `Object.assign` calls. Why do you prefer this style?

File: packages/core/src/browser/source-tree/source-tree.ts [hunk 2]
```diff
@@ -41,50 +43,56 @@ export class SourceTree extends TreeImpl {
...
-            delete updated.expanded;
-            delete updated.children;
+            };
         }
-        if (updated) {
-            if (ExpandableTreeNode.is(updated)) {
-                delete updated.expanded;
+        if (existing) {
+            if (ExpandableTreeNode.is(existing)) {
+                delete (existing as any).expanded;
```
Comment (colin-grant-work) [inline packages/core/src/browser/source-tree/source-tree.ts]: Rather than an `any` cast here, you can use `Partial<ExpandableTreeNode>` (or `CompositeTreeNode`, where appropriate). I'd prefer to avoid file-scoped disabling of `no-explicit-any` where possible.

File: packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts [hunk 3]
```diff
@@ -300,7 +300,7 @@ export class PluginContributionHandler {
         }
 
         if (contributions.colors) {
-            pushContribution('colors', () => this.colors.register(...contributions.colors));
+            pushContribution('colors', () => this.colors.register(...contributions.colors!));
```
Comment (colin-grant-work) [inline packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:303]: Would it be safer to do this:
```typescript
const {colors} = contributions;
pushContribution('colors', () => this.colors.register(...colors));
```
Then TS is happy that colors will exist without the non-null assertion.

File: packages/typehierarchy/src/browser/tree/typehierarchy-tree.ts [hunk 4]
```diff
@@ -31,8 +33,8 @@ export class TypeHierarchyTree extends TreeImpl {
         if (TypeHierarchyTree.Node.is(parent)) {
             await this.ensureResolved(parent);
             if (parent.children.length === 0) {
-                delete parent.children;
-                delete parent.expanded;
+                delete (parent as any).children;
```
Comment (colin-grant-work) [inline packages/typehierarchy/src/browser/tree/typehierarchy-tree.ts:36]: Looks like some `as any` TreeNode casts are still here.

File: packages/core/src/browser/source-tree/source-tree.ts (hunk 1 above)
Comment (paul-marechal) [reply packages/core/src/browser/source-tree/source-tree.ts]: Because TS checks nada when doing `Object.assign`... It's worse that using `any` and relies on useless reflection.

File: packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts (hunk 3 above)
Comment (paul-marechal) [reply packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:303]: I feel like it is OK with the way it currently is, not a fan of binding consts just to please typing...

Comment (colin-grant-work) [reply packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:303]: It isn't _really_ just to please typing in this case. It seems reasonable to believe that nothing will delete the `colors` between this call and when the callback is called, but TS is reminding us that `delete` is a thing, and the rest of this PR shows we're not afraid to use it.

File: packages/core/src/browser/source-tree/source-tree.ts (hunk 1 above)
Comment (colin-grant-work) [reply packages/core/src/browser/source-tree/source-tree.ts]: In that case, I think the `Partial<ExpandableTreeNode>` trick will work here, as well, avoiding the `any`.

File: packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts (hunk 3 above)
Comment (paul-marechal) [reply packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:303]: > It seems reasonable to believe that nothing will delete the colors between this call and when the callback is called [...]

It's the goal of the current implementation.

> [...] but TS is reminding us that `delete` is a thing, and the rest of this PR shows we're not afraid to use it.

Agreed, but once again in this specific instance, adding the `non-null` coercion or capturing the value out of the `contribution` object wouldn't help much should a `delete` be introduced for any reason. In one case it will fail because of an undefined value that was supposed to be defined, and in the other case it will keep working despite the value being deleted.

TL;DR: whenever we have code that isn't specifically tailored like this one, we could do as you say and capture properties to ensure they don't disappear from the object. Although ultimately removing properties mid-flight sounds like a foot gun.

File: packages/core/src/browser/source-tree/source-tree.ts (hunk 1 above)
Comment (paul-marechal) [reply packages/core/src/browser/source-tree/source-tree.ts]: I updated the code with `Partial` like you suggested. We're still kinda lying to the compiler, but without `any`s now.

File: packages/plugin-ext/src/hosted/node/plugin-reader.ts [hunk 5]
```diff
@@ -58,7 +58,8 @@ export class HostedPluginReader implements BackendApplicationContribution {
                         // the request was already closed
                         return;
                     }
-                    if ('code' in e && e['code'] === 'ENOENT') {
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    if ((e as any).code === 'ENOENT') {
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/hosted/node/plugin-reader.ts:62]: If we properly type `e` based on what we might expect then we would not need to cast as `any`?

```ts
res.sendFile(filePath, { root: localPath }, e => {
```

to

```ts
res.sendFile(filePath, { root: localPath }, (e: Error & { code?: string }) => {
```

File: packages/core/src/browser/source-tree/tree-source.ts [hunk 6]
```diff
@@ -37,12 +39,12 @@ export interface CompositeTreeElement extends TreeElement {
     getElements(): MaybePromise<IterableIterator<TreeElement>>
 }
 export namespace CompositeTreeElement {
-    /* eslint-disable @typescript-eslint/no-explicit-any */
-    export function is(element: CompositeTreeElement | any): element is CompositeTreeElement {
-        return !!element && 'getElements' in element;
+    export function is(element?: any): element is CompositeTreeElement {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/source-tree/tree-source.ts:42]: Since we do a `typeof` check, we can simplify it to `unknown`?

File: packages/core/src/browser/tree/tree-expansion.ts [hunk 7]
```diff
@@ -65,15 +67,16 @@ export interface ExpandableTreeNode extends CompositeTreeNode {
 }
 
 export namespace ExpandableTreeNode {
-    export function is(node: Object | undefined): node is ExpandableTreeNode {
-        return !!node && CompositeTreeNode.is(node) && 'expanded' in node;
+    export function is(node?: any): node is ExpandableTreeNode {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-expansion.ts:70]: Can you clarify why we want to now accept `any` given we try to avoid using it project wide?
Why for example in `ExpandableTreeNode` are we not doing like `SelectableTreeNode` and not accepting the parent `TreeNode` as a type?

File: packages/plugin-ext/src/hosted/node/plugin-reader.ts (hunk 5 above)
Comment (paul-marechal) [reply packages/plugin-ext/src/hosted/node/plugin-reader.ts:62]: `any` is more convenient here, we still end up lying to the compiler (can make a typo either in the `{ code?: string }` or `(e as any).code`).

File: packages/core/src/browser/source-tree/tree-source.ts (hunk 6 above)
Comment (paul-marechal) [reply packages/core/src/browser/source-tree/tree-source.ts:42]: Not really since TypeScript makes `object` type guards pretty painful to handle. `any` is fine here.

File: packages/core/src/browser/tree/tree-expansion.ts (hunk 7 above)
Comment (paul-marechal) [reply packages/core/src/browser/tree/tree-expansion.ts:70]: Because it doesn't really matter. The whole point of those functions is to tell TypeScript how to coerce values into specific types: `node is ExpandableTreeNode`.

File: packages/plugin-ext/src/hosted/node/plugin-reader.ts (hunk 5 above)
Comment (paul-marechal) [reply packages/plugin-ext/src/hosted/node/plugin-reader.ts:62]: Found `NodeJS.ErrnoException` that also inherits `Error`, will use that instead.

**Comment by paul-marechal** [comment]
I am contemplating dropping this TypeScript upgrade.

My original idea was that we would be able to use more recent language constructs (named tuple elements, etc...)

The problem I now realize this can cause is that if our typings use those newer constructs, it might break dependents that rely on older versions of TypeScript.

In short, let's keep our old compiler to be as easy to consume as possible?

**Comment by tsmaeder** [comment]
@paul-marechal yet another upgrade policy we should define to be a reliable partner to adopters. The pains of growing up as a project, I guess.
